File: auth/ts/src/grant.ts
/**
 * The grant — docs/auth/grant.md, held to conformance/tables/auth-grant.json.
 *
 * A body in an Archon envelope sealed in `nightseam-grant/1`; a chain of
 * them, root first, held to a request by one pure function; issuance under
 * the issuer's own chain. Every refusal is a code and a hop index; nothing
 * here does I/O, reads a clock or keeps state.
 */
import { getPublicKey } from '@bitspark/archon';
import { open as openEnvelope, seal as sealEnvelope } from '@bitspark/archon-sdk';
import { sha256 } from '@noble/hashes/sha2.js';
import { Reader, compare, concat, equal, fromUtf8, hasControl, u16, u64, utf8, wellFormed } from './bytes.ts';

export const DOMAIN = 'nightseam-grant/1';

/** The bounds of docs/auth/grant.md, by which the work of every function is finite. */
export const MAX_DEPTH = 15;
export const MAX_CHAIN = 16;
export const MAX_ENTRIES = 64;
export const MAX_ENTRY_BYTES = 1024;
export const MAX_BODY_BYTES = 65535;
export const MAX_DOMAIN_BYTES = 255;

const BODY_VERSION = 0x01;

export type Validity = { finite: false } | { finite: true; expiresAt: bigint };

export interface Grant {
  domain: string;
  subject: Uint8Array;
  parent?: Uint8Array;
  scope: string[];
  actions: string[];
  delegable: string[];
  depth: number;
  validity: Validity;
}

export interface Root {
  key: Uint8Array;
  domain: string;
}

export interface Request {
  domain: string;
  action: string;
  scope: string;
}

export type Time = { present: false } | { present: true; now: bigint };

export interface Verified {
  subject: Uint8Array;
  depth: number;
  validity: Validity;
  hops: number;
}

export type Code =
  | 'malformed'
  | 'unsupported_version'
  | 'envelope_invalid'
  | 'chain_too_long'
  | 'domain_mismatch'
  | 'root_mismatch'
  | 'parent_mismatch'
  | 'issuer_mismatch'
  | 'widened_actions'
  | 'widened_scope'
  | 'widened_depth'
  | 'widened_validity'
  | 'not_covered'
  | 'time_required'
  | 'expired';

export const CODES: readonly Code[] = [
  'malformed',
  'unsupported_version',
  'envelope_invalid',
  'chain_too_long',
  'domain_mismatch',
  'root_mismatch',
  'parent_mismatch',
  'issuer_mismatch',
  'widened_actions',
  'widened_scope',
  'widened_depth',
  'widened_validity',
  'not_covered',
  'time_required',
  'expired',
];

export interface Refusal {
  code: Code;
  hop: number;
}

export function isRefusal(value: Verified | Refusal): value is Refusal {
  return 'code' in value;
}

/** What `encode` throws for a body outside the rules: a code, never prose about the body. */
export class GrantError extends Error {
  readonly code: Code;
  constructor(code: Code) {
    super(code);
    this.name = 'GrantError';
    this.code = code;
  }
}

// ---- the body ------------------------------------------------------------------------

/** A domain or an entry: UTF-8, no control character, within its byte bound. */
function textBytes(text: unknown, max: number): Uint8Array | undefined {
  if (typeof text !== 'string' || !wellFormed(text) || hasControl(text)) return undefined;
  const bytes = utf8(text);
  if (bytes.length === 0 || bytes.length > max) return undefined;
  return bytes;
}

/** A list of entries: at most 64, each within bound, strictly ascending by bytes. */
function listBytes(entries: unknown): Uint8Array[] | undefined {
  if (!Array.isArray(entries) || entries.length > MAX_ENTRIES) return undefined;
  const out: Uint8Array[] = [];
  for (const entry of entries) {
    const bytes = textBytes(entry, MAX_ENTRY_BYTES);
    if (bytes === undefined) return undefined;
    if (out.length > 0 && compare(out[out.length - 1]!, bytes) >= 0) return undefined;
    out.push(bytes);
  }
  return out;
}

function encodeList(entries: Uint8Array[]): Uint8Array {
  const parts: Uint8Array[] = [u16(entries.length)];
  for (const e of entries) parts.push(u16(e.length), e);
  return concat(...parts);
}

function within(inner: Uint8Array[], outer: Uint8Array[]): boolean {
  return inner.every((e) => outer.some((o) => equal(e, o)));
}

/** `a/b` covers `a/b` and everything under `a/b/`. */
export function covers(entry: string, scope: string): boolean {
  return scope === entry || scope.startsWith(entry + '/');
}

function validValidity(v: unknown): v is Validity {
  if (typeof v !== 'object' || v === null) return false;
  const x = v as { finite?: unknown; expiresAt?: unknown };
  if (x.finite === false) return true;
  return (
    x.finite === true && typeof x.expiresAt === 'bigint' && x.expiresAt >= 0n && x.expiresAt <= 0xffffffffffffffffn
  );
}

/** The one encoding of a grant; throws `GrantError('malformed')` for a body outside the rules. */
export function encode(g: Grant): Uint8Array {
  const domain = textBytes(g.domain, MAX_DOMAIN_BYTES);
  const scope = listBytes(g.scope);
  const actions = listBytes(g.actions);
  const delegable = listBytes(g.delegable);
  if (
    domain === undefined ||
    scope === undefined ||
    actions === undefined ||
    delegable === undefined ||
    !(g.subject instanceof Uint8Array) ||
    g.subject.length !== 32 ||
    (g.parent !== undefined && (!(g.parent instanceof Uint8Array) || g.parent.length !== 32)) ||
    !Number.isInteger(g.depth) ||
    g.depth < 0 ||
    g.depth > MAX_DEPTH ||
    !validValidity(g.validity) ||
    !within(delegable, actions)
  ) {
    throw new GrantError('malformed');
  }
  const body = concat(
    new Uint8Array([BODY_VERSION, domain.length]),
    domain,
    g.subject,
    g.parent === undefined ? new Uint8Array([0x00]) : concat(new Uint8Array([0x01]), g.parent),
    encodeList(scope),
    encodeList(actions),
    encodeList(delegable),
    new Uint8Array([g.depth]),
    g.validity.finite ? concat(new Uint8Array([0x01]), u64(g.validity.expiresAt)) : new Uint8Array([0x00]),
  );
  if (body.length > MAX_BODY_BYTES) throw new GrantError('malformed');
  return body;
}

function readList(r: Reader): string[] | undefined {
  const count = r.u16();
  if (count === undefined || count > MAX_ENTRIES) return undefined;
  const out: string[] = [];
  let previous: Uint8Array | undefined;
  for (let i = 0; i < count; i++) {
    const n = r.u16();
    if (n === undefined || n === 0 || n > MAX_ENTRY_BYTES) return undefined;
    const bytes = r.take(n);
    if (bytes === undefined) return undefined;
    const text = fromUtf8(bytes);
    if (text === undefined || hasControl(text)) return undefined;
    if (previous !== undefined && compare(previous, bytes) >= 0) return undefined;
    previous = bytes;
    out.push(text);
  }
  return out;
}

/** Total: a grant, or the code that says why these bytes are not one. */
export function decode(body: Uint8Array): Grant | Code {
  if (body.length === 0 || body.length > MAX_BODY_BYTES) return 'malformed';
  if (body[0] !== BODY_VERSION) return 'unsupported_version';
  const r = new Reader(body);
  r.u8();
  const dlen = r.u8();
  if (dlen === undefined || dlen === 0) return 'malformed';
  const domainBytes = r.take(dlen);
  if (domainBytes === undefined) return 'malformed';
  const domain = fromUtf8(domainBytes);
  if (domain === undefined || hasControl(domain)) return 'malformed';
  const subject = r.take(32);
  if (subject === undefined) return 'malformed';
  const parentKind = r.u8();
  let parent: Uint8Array | undefined;
  if (parentKind === 0x01) {
    parent = r.take(32);
    if (parent === undefined) return 'malformed';
  } else if (parentKind !== 0x00) {
    return 'malformed';
  }
  const scope = readList(r);
  const actions = readList(r);
  const delegable = readList(r);
  if (scope === undefined || actions === undefined || delegable === undefined) return 'malformed';
  if (!delegable.every((d) => actions.includes(d))) return 'malformed';
  const depth = r.u8();
  if (depth === undefined || depth > MAX_DEPTH) return 'malformed';
  const validityKind = r.u8();
  let validity: Validity;
  if (validityKind === 0x00) {
    validity = { finite: false };
  } else if (validityKind === 0x01) {
    const expiresAt = r.u64();
    if (expiresAt === undefined) return 'malformed';
    validity = { finite: true, expiresAt };
  } else {
    return 'malformed';
  }
  if (r.remaining !== 0) return 'malformed';
  return { domain, subject, parent, scope, actions, delegable, depth, validity };
}

// ---- the envelope --------------------------------------------------------------------

/** Seals the encoded body by the issuer's key in the grant domain. */
export function seal(issuerSeed: Uint8Array, g: Grant): Uint8Array {
  return sealEnvelope(issuerSeed, DOMAIN, encode(g));
}

/** Opens an envelope in the grant domain: the issuer and the grant, or the code. */
export function open(envelope: Uint8Array): { issuer: Uint8Array; grant: Grant } | Code {
  let opened: { pubkey: Uint8Array; payload: Uint8Array };
  try {
    opened = openEnvelope(envelope, DOMAIN);
  } catch {
    return 'envelope_invalid';
  }
  const grant = decode(opened.payload);
  if (typeof grant === 'string') return grant;
  return { issuer: opened.pubkey, grant };
}

/** The SHA-256 of the whole envelope: what a child names as its parent. */
export function digest(envelope: Uint8Array): Uint8Array {
  return sha256(envelope);
}

// ---- the chain ------------------------------------------------------------------------

interface Hop {
  issuer: Uint8Array;
  grant: Grant;
  digest: Uint8Array;
}

/** Steps 1 and 2 of the evaluation: length, then each hop's envelope, ancestry, domain and attenuation. */
function openChain(root: Root, chain: Uint8Array[]): Hop[] | Refusal {
  if (chain.length === 0) return { code: 'malformed', hop: 0 };
  if (chain.length > MAX_CHAIN) return { code: 'chain_too_long', hop: 0 };
  const hops: Hop[] = [];
  for (let i = 0; i < chain.length; i++) {
    const opened = open(chain[i]!);
    if (typeof opened === 'string') return { code: opened, hop: i };
    const { issuer, grant } = opened;
    const previous = hops[i - 1];
    if (previous === undefined) {
      if (grant.parent !== undefined) return { code: 'parent_mismatch', hop: i };
      if (!equal(issuer, root.key)) return { code: 'root_mismatch', hop: i };
    } else {
      if (grant.parent === undefined || !equal(grant.parent, previous.digest))
        return { code: 'parent_mismatch', hop: i };
      if (!equal(issuer, previous.grant.subject)) return { code: 'issuer_mismatch', hop: i };
    }
    if (grant.domain !== root.domain) return { code: 'domain_mismatch', hop: i };
    if (previous !== undefined) {
      const widened = attenuates(previous.grant, grant);
      if (widened !== undefined) return { code: widened, hop: i };
    }
    hops.push({ issuer, grant, digest: digest(chain[i]!) });
  }
  return hops;
}

/** The attenuation rule between a parent and its child, in the packet's order. */
function attenuates(parent: Grant, child: Grant): Code | undefined {
  if (!(parent.depth > 0 && child.depth < parent.depth)) return 'widened_depth';
  if (!child.actions.every((a) => parent.delegable.includes(a))) return 'widened_actions';
  if (!child.scope.every((s) => parent.scope.some((p) => covers(p, s)))) return 'widened_scope';
  if (parent.validity.finite) {
    if (!child.validity.finite || child.validity.expiresAt > parent.validity.expiresAt) return 'widened_validity';
  }
  return undefined;
}

/** Step 3: the request covered at every hop, delegable at every hop but the last. */
function coverage(root: Root, hops: Hop[], request: Request): Refusal | undefined {
  if (request.domain !== root.domain) return { code: 'domain_mismatch', hop: 0 };
  for (let i = 0; i < hops.length; i++) {
    const g = hops[i]!.grant;
    const last = i === hops.length - 1;
    if (!g.actions.includes(request.action)) return { code: 'not_covered', hop: i };
    if (!g.scope.some((s) => covers(s, request.scope))) return { code: 'not_covered', hop: i };
    if (!last && !g.delegable.includes(request.action)) return { code: 'not_covered', hop: i };
  }
  return undefined;
}

/** Step 4: one decision time, required where any hop is finite, strictly before each finite expiry. */
function time(hops: Hop[], now: Time): Refusal | undefined {
  const first = hops.findIndex((h) => h.grant.validity.finite);
  if (first < 0) return undefined;
  if (!now.present) return { code: 'time_required', hop: first };
  for (let i = 0; i < hops.length; i++) {
    const v = hops[i]!.grant.validity;
    if (v.finite && now.now >= v.expiresAt) return { code: 'expired', hop: i };
  }
  return undefined;
}

function verified(hops: Hop[]): Verified {
  const leaf = hops[hops.length - 1]!.grant;
  let validity: Validity = { finite: false };
  for (const h of hops) {
    const v = h.grant.validity;
    if (v.finite && (!validity.finite || v.expiresAt < validity.expiresAt))
      validity = { finite: true, expiresAt: v.expiresAt };
  }
  return { subject: leaf.subject, depth: leaf.depth, validity, hops: hops.length };
}

/** Steps 1, 2 and 4: the chain held without a request. */
export function inspect(root: Root, chain: Uint8Array[], now: Time): Verified | Refusal {
  const hops = openChain(root, chain);
  if (!Array.isArray(hops)) return hops;
  const late = time(hops, now);
  if (late !== undefined) return late;
  return verified(hops);
}

/** The whole evaluation: the chain held to the request at this time. */
export function verify(root: Root, chain: Uint8Array[], request: Request, now: Time): Verified | Refusal {
  const hops = openChain(root, chain);
  if (!Array.isArray(hops)) return hops;
  const uncovered = coverage(root, hops, request);
  if (uncovered !== undefined) return uncovered;
  const late = time(hops, now);
  if (late !== undefined) return late;
  return verified(hops);
}

// ---- issuance -------------------------------------------------------------------------

/**
 * Signs a child under the issuer's own chain — root first, ending in a grant
 * to the issuer — after holding that chain as `inspect` does and the child to
 * the attenuation rule against its leaf. The child's parent is the leaf's
 * digest, set here; with `inherit` its validity resolves to the leaf's before
 * signing. A root grant has an empty chain, is issued by the root key alone
 * and states its validity. Deterministic: the same inputs seal the same bytes.
 */
export function issue(
  root: Root,
  parents: Uint8Array[],
  issuerSeed: Uint8Array,
  child: Grant,
  inherit: boolean,
  now: Time,
): Uint8Array | Refusal {
  let issuerKey: Uint8Array;
  try {
    issuerKey = getPublicKey(issuerSeed);
  } catch {
    return { code: 'malformed', hop: 0 };
  }
  const hop = parents.length;
  let grant: Grant;
  if (parents.length === 0) {
    if (inherit) return { code: 'malformed', hop: 0 };
    if (child.parent !== undefined) return { code: 'parent_mismatch', hop: 0 };
    if (!equal(issuerKey, root.key)) return { code: 'root_mismatch', hop: 0 };
    grant = { ...child };
  } else {
    const hops = openChain(root, parents);
    if (!Array.isArray(hops)) return hops;
    const late = time(hops, now);
    if (late !== undefined) return late;
    const leaf = hops[hops.length - 1]!;
    if (!equal(issuerKey, leaf.grant.subject)) return { code: 'issuer_mismatch', hop };
    if (child.parent !== undefined && !equal(child.parent, leaf.digest)) return { code: 'parent_mismatch', hop };
    grant = { ...child, parent: leaf.digest, validity: inherit ? leaf.grant.validity : child.validity };
    const widened = attenuates(leaf.grant, grant);
    if (widened !== undefined) return { code: widened, hop };
  }
  try {
    return seal(issuerSeed, grant);
  } catch (error) {
    if (error instanceof GrantError) return { code: error.code, hop };
    return { code: 'malformed', hop };
  }
}

File: auth/go/grant/grant.go
package grant

import (
	"encoding/binary"
	"errors"
	"fmt"
	"unicode/utf8"
)

// Domain is the signing domain every grant envelope is sealed in. A
// signature made in any other domain, or raw, verifies as nothing here, and
// a grant signature can never be replayed as a possession proof or as any
// other envelope of the profile.
const Domain = "nightseam-grant/1"

// The bounds of docs/auth/grant.md, by which every piece of work here is
// linear in its input and refused before it grows.
const (
	version        byte = 0x01
	maxDepth            = 15
	maxChain            = 16
	maxEntries          = 64
	maxEntryBytes       = 1024
	maxBodyBytes        = 65535
	maxDomainBytes      = 255
)

// Code is a refusal's code: the closed set a program branches on. A refusal
// echoes no member of any grant; it is a code and a hop and nothing else.
type Code string

// The codes, in the order the evaluation can produce them.
const (
	Malformed          Code = "malformed"
	UnsupportedVersion Code = "unsupported_version"
	EnvelopeInvalid    Code = "envelope_invalid"
	ChainTooLong       Code = "chain_too_long"
	DomainMismatch     Code = "domain_mismatch"
	RootMismatch       Code = "root_mismatch"
	ParentMismatch     Code = "parent_mismatch"
	IssuerMismatch     Code = "issuer_mismatch"
	WidenedActions     Code = "widened_actions"
	WidenedScope       Code = "widened_scope"
	WidenedDepth       Code = "widened_depth"
	WidenedValidity    Code = "widened_validity"
	NotCovered         Code = "not_covered"
	TimeRequired       Code = "time_required"
	Expired            Code = "expired"
)

// Validity is a grant's explicit validity: unbounded, or finite with an
// absolute expiry in seconds since the Unix epoch, UTC. Nothing is an
// implicit infinity; an unbounded validity carries no expiry.
type Validity struct {
	Finite    bool
	ExpiresAt uint64
}

// Grant is the body of a grant: what a grant means. Who said it is the
// envelope's public key, which the body does not repeat.
type Grant struct {
	Domain    string    // the trust domain the scope lives in
	Subject   [32]byte  // the Ed25519 public key the grant is issued to
	Parent    *[32]byte // nil for a root grant; else the SHA-256 of the parent's whole envelope
	Scope     []string  // what the subject may reach, strictly ascending, at most 64
	Actions   []string  // what the subject may do there, the same rules
	Delegable []string  // which of those it may pass on: a subset of Actions
	Depth     uint8     // how many further hops it may pass them, 0..=15
	Validity  Validity
}

// Root is what a verifier is configured with: one public key and one trust
// domain.
type Root struct {
	Key    [32]byte
	Domain string
}

// Request is one concrete request a chain covers at every hop or covers
// nothing.
type Request struct {
	Domain, Action, Scope string
}

// Time is the decision time: absent, or a number of seconds since the Unix
// epoch, UTC. It is the caller's; nothing here reads a clock.
type Time struct {
	Present bool
	Now     uint64
}

// Verified is what a chain that holds yields: the leaf's subject, its
// depth, the chain's effective validity — the earliest finite expiry, or
// unbounded — and the hop count.
type Verified struct {
	Subject  [32]byte
	Depth    uint8
	Validity Validity
	Hops     int
}

// Refusal is a code and the index of the hop it arose at, and nothing
// else.
type Refusal struct {
	Code Code
	Hop  int
}

// Encode is the one canonical encoding of a body: version-tagged, fixed
// order, length-prefixed, big-endian. It refuses a body outside the rules
// of the packet — a domain or entry that is empty, too long, not UTF-8 or
// carrying a control character; a list out of strictly ascending byte
// order, with a duplicate, or of more than 64 entries; a delegable outside
// the actions; a depth above 15; an unbounded validity carrying an expiry;
// a body over 65535 bytes.
func Encode(g Grant) ([]byte, error) {
	if err := check(g); err != nil {
		return nil, err
	}
	body := make([]byte, 0, 2+len(g.Domain)+32+33+8+2)
	body = append(body, version)
	body = append(body, byte(len(g.Domain)))
	body = append(body, g.Domain...)
	body = append(body, g.Subject[:]...)
	if g.Parent == nil {
		body = append(body, 0x00)
	} else {
		body = append(body, 0x01)
		body = append(body, g.Parent[:]...)
	}
	body = appendList(body, g.Scope)
	body = appendList(body, g.Actions)
	body = appendList(body, g.Delegable)
	body = append(body, g.Depth)
	if g.Validity.Finite {
		body = append(body, 0x01)
		body = binary.BigEndian.AppendUint64(body, g.Validity.ExpiresAt)
	} else {
		body = append(body, 0x00)
	}
	if len(body) > maxBodyBytes {
		return nil, fmt.Errorf("grant: body is %d bytes, over the bound of %d", len(body), maxBodyBytes)
	}
	return body, nil
}

func appendList(body []byte, entries []string) []byte {
	body = binary.BigEndian.AppendUint16(body, uint16(len(entries)))
	for _, entry := range entries {
		body = binary.BigEndian.AppendUint16(body, uint16(len(entry)))
		body = append(body, entry...)
	}
	return body
}

// Decode is total: it reads a body and answers a grant or a code, never a
// panic. A first byte other than the version is UnsupportedVersion; an
// empty body, a body over 65535 bytes, a truncated field, a trailing byte,
// an unknown parent or validity kind, a list out of order or over its
// bounds, and a delegable outside the actions are Malformed.
func Decode(body []byte) (Grant, Code) {
	if len(body) == 0 {
		return Grant{}, Malformed
	}
	if body[0] != version {
		return Grant{}, UnsupportedVersion
	}
	if len(body) > maxBodyBytes {
		return Grant{}, Malformed
	}
	r := reader{rest: body[1:]}
	var g Grant
	domainLen, ok := r.u8()
	if !ok {
		return Grant{}, Malformed
	}
	domain, ok := r.take(int(domainLen))
	if !ok {
		return Grant{}, Malformed
	}
	g.Domain = string(domain)
	subject, ok := r.take(32)
	if !ok {
		return Grant{}, Malformed
	}
	copy(g.Subject[:], subject)
	parentKind, ok := r.u8()
	if !ok {
		return Grant{}, Malformed
	}
	switch parentKind {
	case 0x00:
	case 0x01:
		parent, ok := r.take(32)
		if !ok {
			return Grant{}, Malformed
		}
		g.Parent = new([32]byte)
		copy(g.Parent[:], parent)
	default:
		return Grant{}, Malformed
	}
	if g.Scope, ok = r.list(); !ok {
		return Grant{}, Malformed
	}
	if g.Actions, ok = r.list(); !ok {
		return Grant{}, Malformed
	}
	if g.Delegable, ok = r.list(); !ok {
		return Grant{}, Malformed
	}
	if g.Depth, ok = r.u8(); !ok {
		return Grant{}, Malformed
	}
	validityKind, ok := r.u8()
	if !ok {
		return Grant{}, Malformed
	}
	switch validityKind {
	case 0x00:
	case 0x01:
		expiresAt, ok := r.take(8)
		if !ok {
			return Grant{}, Malformed
		}
		g.Validity = Validity{Finite: true, ExpiresAt: binary.BigEndian.Uint64(expiresAt)}
	default:
		return Grant{}, Malformed
	}
	if len(r.rest) != 0 {
		return Grant{}, Malformed
	}
	if check(g) != nil {
		return Grant{}, Malformed
	}
	return g, ""
}

// reader consumes a body front to back; every read reports whether the
// bytes were there.
type reader struct {
	rest []byte
}

func (r *reader) u8() (byte, bool) {
	if len(r.rest) < 1 {
		return 0, false
	}
	b := r.rest[0]
	r.rest = r.rest[1:]
	return b, true
}

func (r *reader) u16() (uint16, bool) {
	if len(r.rest) < 2 {
		return 0, false
	}
	n := binary.BigEndian.Uint16(r.rest)
	r.rest = r.rest[2:]
	return n, true
}

func (r *reader) take(n int) ([]byte, bool) {
	if len(r.rest) < n {
		return nil, false
	}
	b := r.rest[:n]
	r.rest = r.rest[n:]
	return b, true
}

// list reads u16(count) ‖ (u16(len) ‖ entry)*, refusing a count or an entry
// over its bound before reading it, so the work is bounded by the packet
// and not by the input.
func (r *reader) list() ([]string, bool) {
	count, ok := r.u16()
	if !ok || count > maxEntries {
		return nil, false
	}
	if count == 0 {
		return nil, true
	}
	entries := make([]string, 0, count)
	for i := 0; i < int(count); i++ {
		n, ok := r.u16()
		if !ok || n > maxEntryBytes {
			return nil, false
		}
		entry, ok := r.take(int(n))
		if !ok {
			return nil, false
		}
		entries = append(entries, string(entry))
	}
	return entries, true
}

// check holds a grant to the rules of the body; Encode refuses what it
// refuses and Decode calls it Malformed.
func check(g Grant) error {
	if err := text(g.Domain, maxDomainBytes); err != nil {
		return fmt.Errorf("grant: domain: %w", err)
	}
	for _, list := range []struct {
		name    string
		entries []string
	}{{"scope", g.Scope}, {"actions", g.Actions}, {"delegable", g.Delegable}} {
		if err := checkList(list.entries); err != nil {
			return fmt.Errorf("grant: %s: %w", list.name, err)
		}
	}
	if !subset(g.Delegable, g.Actions) {
		return errors.New("grant: delegable is not a subset of actions")
	}
	if g.Depth > maxDepth {
		return fmt.Errorf("grant: depth %d is above the bound of %d", g.Depth, maxDepth)
	}
	if !g.Validity.Finite && g.Validity.ExpiresAt != 0 {
		return errors.New("grant: an unbounded validity carries no expiry")
	}
	return nil
}

// checkList holds a list to its rules: at most 64 entries, each 1..=1024
// bytes of UTF-8 with no control character, in strictly ascending byte
// order — canonical and duplicate-free by construction.
func checkList(entries []string) error {
	if len(entries) > maxEntries {
		return fmt.Errorf("%d entries, over the bound of %d", len(entries), maxEntries)
	}
	for i, entry := range entries {
		if err := text(entry, maxEntryBytes); err != nil {
			return fmt.Errorf("entry %d: %w", i, err)
		}
		if i > 0 && entries[i-1] >= entry {
			return fmt.Errorf("entry %d is not strictly after entry %d", i, i-1)
		}
	}
	return nil
}

// text holds a domain or an entry to the shared rule: 1..=max bytes of
// valid UTF-8 with no control character (U+0000–U+001F, U+007F).
func text(s string, max int) error {
	if len(s) == 0 || len(s) > max {
		return fmt.Errorf("%d bytes, want 1..=%d", len(s), max)
	}
	if !utf8.ValidString(s) {
		return errors.New("not valid UTF-8")
	}
	for _, r := range s {
		if r < 0x20 || r == 0x7f {
			return fmt.Errorf("carries the control character U+%04X", r)
		}
	}
	return nil
}

// subset reports whether every entry of inner is an entry of outer; both
// are sorted, so one walk decides it.
func subset(inner, outer []string) bool {
	j := 0
	for _, entry := range inner {
		for j < len(outer) && outer[j] < entry {
			j++
		}
		if j == len(outer) || outer[j] != entry {
			return false
		}
	}
	return true
}

// contains reports whether entries, sorted, holds entry.
func contains(entries []string, entry string) bool {
	for _, e := range entries {
		if e == entry {
			return true
		}
		if e > entry {
			return false
		}
	}
	return false
}

// covered reports whether one of the entries covers scope: an entry is a
// path, and a/b covers a/b and everything under a/b/ — never a/bc.
func covered(scope string, entries []string) bool {
	for _, entry := range entries {
		if scope == entry {
			return true
		}
		if len(scope) > len(entry) && scope[:len(entry)] == entry && scope[len(entry)] == '/' {
			return true
		}
	}
	return false
}

// allCovered reports whether every entry of inner is covered by an entry of
// outer.
func allCovered(inner, outer []string) bool {
	for _, entry := range inner {
		if !covered(entry, outer) {
			return false
		}
	}
	return true
}
